yt/data_objects/tests/test_firefly.py

```diff
@@ -36,3 +42,127 @@ def test_firefly_write_to_disk():
     )
 
     reader.dumpToJSON()
+
+
+@pytest.fixture
+def firefly_test_dataset():
+    # create dataset
+    ds_fields = [
+        # Assumed present
+        ("PartType1", "particle_position_x"),
+        ("PartType1", "particle_position_y"),
+        ("PartType1", "particle_position_z"),
+        ("gas", "particle_position_x"),
+        ("gas", "particle_position_y"),
+        ("gas", "particle_position_z"),
+        # User input
+        ("PartType1", "Masses"),
+        ("gas", "Masses"),
+        ("gas", "Temperature"),
+    ]
+    ds_field_units = [
+        "code_length",
+        "code_length",
+        "code_length",
+        "code_length",
+        "code_length",
+        "code_length",
+        "code_mass",
+        "code_mass",
+        "code_temperature",
+    ]
+    ds_negative = [
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+    ]
+    ds = fake_particle_ds(
+        fields=ds_fields,
+        units=ds_field_units,
+        negative=ds_negative,
+    )
+    dd = ds.all_data()
+    return (ds, dd)
+
+
+@requires_module("firefly")
+@pytest.mark.parametrize(
+    "fields_to_include,fields_units,pgs_to_test,expectation",
+    [
+        (None, None, None, dnr()),  # Test default values
+        ([], [], None, dnr()),  # Test empty fields
+        (["Masses"], ["code_mass"], None, dnr()),  # Test common field (Masses)
+        (
+            ["Temperature"],
+            ["code_temperature"],
+            "gas",
+            dnr(),
+        ),  # Test unique field (Temperature)
+        (
+            ["dinos"],
+            ["code_length"],
+            None,
+            pytest.raises(YTFieldNotFound),
```

Comment:
So I don't think I've ever seen this done this way, and I am uncertain that this is supported, because `pytest.raises` and `pytest.warns` are context managers, and as such they *need* to be initialised through a `with` statement, so I doubt that you can parametrize your expectations this way.
Alternatively, you may parametrize the exception type.

Reply:
I pulled this type of construction directly from the [pytest documentation](https://docs.pytest.org/en/6.2.x/example/parametrize.html#parametrizing-conditional-raising), so I think it's right? The other option I thought of was using 
`pytest.param(["dinos"], ["code_length"], None, mark=pytest.mark.xfail)`
and likewise for the nonexistent tuple, but it seemed like xfails were more for non-implemented features/unfixed bugs?